mediagrains/utils/h264_parser.py

```diff
@@ -0,0 +1,700 @@
+                pps = self._parse_pps(rbsp_bits)
+                self.pps_sets[pps.pic_parameter_set_id] = pps
+
+            offset = next_offset
+
+        if primary_slice_headers:
+            return self._extract_frame_info(primary_slice_headers)
+        else:
+            return None
+
+    def _extract_frame_info(self,
+                            primary_slice_headers: List[SliceHeader]) -> Optional[FrameInfo]:
+        """Extract frame info from the parsed SPS
+
+        :param primary_slice_headers: the primary slice headers
+        :returns: a FrameInfo if the active SPS is available, else None
+        """
+        if not primary_slice_headers:
+            return None
+
+        # Get the active SPS
+        if primary_slice_headers[0].pic_parameter_set_id in self.pps_sets:
+            pps_id = primary_slice_headers[0].pic_parameter_set_id
+            pps = self.pps_sets[pps_id]
+            if pps.seq_parameter_set_id in self.sps_sets:
+                sps = self.sps_sets[pps.seq_parameter_set_id]
+        if not sps:
+            return None
+
+        frame_info = FrameInfo()
+
+        frame_info.profile = sps.profile_idc
+        frame_info.flags = sps.flags
+        frame_info.level = sps.level_idc
+        if sps.bit_depth_luma_minus8 is not None:
+            frame_info.bit_depth = sps.bit_depth_luma_minus8 + 8
+
+        frame_info.coded_width = (sps.pic_width_in_mbs_minus1 + 1) * 16
+        frame_info.coded_height = (sps.pic_height_in_map_units_minus1 + 1) * 16 * (2 - sps.frame_mbs_only_flag)
+
+        if sps.chroma_array_type == 0:
+            crop_unit_x = 1
+            crop_unit_y = 2 - sps.frame_mbs_only_flag
+        else:
+            if sps.chroma_array_type == 1:
+                sub_width_c = 2
+                sub_height_c = 2
+            elif sps.chroma_array_type == 2:
+                sub_width_c = 2
+                sub_height_c = 1
+            else:  # sps.chroma_array_type == 3
+                sub_width_c = 1
+                sub_height_c = 1
+            crop_unit_x = sub_width_c
+            crop_unit_y = sub_height_c * (2 - sps.frame_mbs_only_flag)
+
+        x_offset = crop_unit_x * sps.frame_crop_left_offset
+        y_offset = crop_unit_y * sps.frame_crop_top_offset
+
+        frame_info.width = frame_info.coded_width - crop_unit_x * sps.frame_crop_right_offset - x_offset
+        frame_info.height = frame_info.coded_height - crop_unit_y * sps.frame_crop_bottom_offset - y_offset
+
+        frame_info.sample_aspect_ratio = self._get_sample_aspect_ratio(sps)
+
+        if sps.timing_info_present_flag:
+            if sps.time_scale == 0 or sps.num_units_in_tick == 0:
+                raise ValueError("Invalid timing info: time_scale={}, num_units_in_tick={}".format(
+                    sps.time_scale, sps.num_units_in_tick
+                ))
+
+            # TODO: implement D.2.2 of H.264 spec. and explanation surrounding Table E-6
```

Comment:
Does this TODO need resolving?

Reply:
This was a TODO in the C++ code and the reason is it has been implemented is because it is a lot of complexity for probably little practical use. I'll change this TODO comment into a NOTE.